testinfra/backend/base.py

```diff
@@ -301,6 +256,36 @@ def encode(self, data: str) -> bytes:
             return data.encode(self.encoding)
 
     def result(self, *args: Any, **kwargs: Any) -> CommandResult:
-        result = CommandResult(self, *args, **kwargs)
+        rc, cmd = args[:2]
+
+        out_bytes, err_bytes = b"", b""
+        if len(args) > 2:
+            out_bytes = args[2]
+        if len(args) > 3:
+            err_bytes = args[3]
+
+        out_bytes = kwargs.get("stdout_bytes", out_bytes)
+        err_bytes = kwargs.get("stderr_bytes", err_bytes)
+
+        out = kwargs.get("stdout", "")
+        if not out and out_bytes:
+            out = self.decode(out_bytes)
+        elif out and not out_bytes:
+            out_bytes = self.encode(out)
```

Comment:
I think we should keep lazy encoding or decoding, i.e. using properties, because sometime we may want to get binary data which cannot be converted to text.

I think the dataclass should have `_stdout` and `_stdout_bytes` and check at `__init__` that at least one of them are set. Or make subclasses BinaryCommandResult and TextCommandResult.

Reply:
Why is lazy encoding and decoding are needed? My way of thoughts here was to remove dependency on `Backend`, which looks redundant to me. So why is it needed here?

With `dataclass` implementation we always know that parameters are set (at least as `None`, but that is the task for `mypy` to detect), so this verification is kinda included.

Current solution doesn't break backward compatibility, at least to the extend it is tested now.

Another solution here would be to do similar to what `subprocess.CompletedProcess` does: it is generic and can be either `bytes` of `str`. We can borrow the approach, but that will change the API.

Reply:
Lazy encoding/decoding is needed because commands like `run("cat /usr/bin/cat")` will raise an error if trying to decode. Also IIRC either salt and/or ansible might have weird encoding behavior while returning command output in a json field. Lazy encoding/decoding avoid some error, as long as user doesn't explicitly ask for encoding/decoding.

Yes something like this could be fine:

```python
@dataclass
class CommandResult:
    _stdout: AnyStr
    _stderr: AnyStr

    @property
    def stdout(self) -> str:
        if isinstance(self._stdout, bytes):
            return self._stdout.encode(...)
        return self._stdout

    [...]
```

Reply:
_(took longer than I expected, I was busy)_

Applied changes to keep lazy encoding/decoding.
@philpep do you think it is time to change `result(self, *args: Any, **kwargs: Any)` signature in this PR?